src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,283 @@
+   * server. Based on the last row key calculated by data already passed through
+   * the chunk transformer, the set of row keys and row ranges is calculated and
+   * updated. The calculated row keys and ranges are used along with other
+   * properties provided by the user like limits and filters to compute and
+   * return a request that will be used in the next read rows call.
+   *
+   * @return {protos.google.bigtable.v2.IReadRowsRequest} The request options
+   * for the next readrows request.
+   */
+  getResumeRequest(): protos.google.bigtable.v2.IReadRowsRequest {
+    this.updateKeysAndRanges();
+    const reqOpts = this
+      .tableStrategyInfo as google.bigtable.v2.IReadRowsRequest;
+
+    // Create the new reqOpts
+    reqOpts.rows = {};
+
+    // Preprocess all the keys and ranges to Bytes
+    reqOpts.rows.rowKeys = this.rowKeys.map(
+      Mutation.convertToBytes
+    ) as {} as Uint8Array[];
+
+    reqOpts.rows.rowRanges = this.ranges.map(range =>
+      Filter.createRange(
+        range.start as BoundData,
+        range.end as BoundData,
+        'Key'
+      )
+    );
+
+    if (this.options.filter) {
+      reqOpts.filter = Filter.parse(this.options.filter);
+    }
+
+    if (this.hasLimit) {
+      reqOpts.rowsLimit = this.rowsLimit - this.rowsRead;
+    }
+    return reqOpts;
+  }
+
+  /**
+   * Decides if the client is going to retry a request.
+   *
+   * canResume contains the logic that will decide if the client will retry with
+   * another request when it receives an error. This logic is passed along to
+   * google-gax and used by google-gax to decide if the client should retry
+   * a request when google-gax receives an error. If canResume returns true then
+   * the client will retry with another request computed by getResumeRequest. If
+   * canResume request returns false then the error will bubble up from gax to
+   * the handwritten layer.
+   *
+   * @param {GoogleError} [error] The error that Google Gax receives.
+   * @return {boolean} True if the client will retry
+   */
+  canResume(error: GoogleError): boolean {
+    // First update the row keys and the row ranges based on the last row key.
+    this.updateKeysAndRanges();
+    if (error.statusDetails === 'RetryInfo') {
+      return true;
+    }
+    // If all the row keys and ranges are read, end the stream
+    // and do not retry.
+    if (this.rowKeys.length === 0 && this.ranges.length === 0) {
+      return false;
+    }
+    // If there was a row limit in the original request and
+    // we've already read all the rows and met/exceeded that limit, end the
+    // stream and do not retry.
+    if (this.hasLimit && this.rowsLimit === this.rowsRead) {
+      return false;
+    }
+    const retryCodesUsed = this.retryCodes
+      ? this.retryCodes
+      : arrify(RETRYABLE_STATUS_CODES);
+    if (
+      error.code &&
+      (retryCodesUsed.includes(error.code) || isRstStreamError(error))
```

Comment:
I am not sure what to do about this. Currently `isRstStreamError` checks if the error code is 13 and the message has RST stream in here because that is what the client library did before. If the user provides 13 in a custom set of codes then should it always retry? If they don't provide 13 in a custom set of codes then should it never retry on code 13 or should it only retry on RST stream errors? If we allow the user to override `canResume` then they will have control over this behaviour.

Longer term, I think this retry logic should be generated code that is the same across client library languages at some point.

Reply:
Yeah, I'm not really sure what a good solution is either, just wanted to make sure you were aware of the situation, and maybe discussed it with stakeholders. 

It still seems strange to me to retry based on the content of the error, instead of just the error code itself